pvlib/ivtools.py

```diff
@@ -0,0 +1,304 @@
+    and more points increase the likelihood of extracting reasonable
+    parameters.
+
+    References
+    ----------
+    [1] S.R. Wenham, M.A. Green, M.E. Watt, "Applied Photovoltaics" ISBN
+    0 86758 909 4
+    [2] C. B. Jones, C. W. Hansen, Single Diode Parameter Extraction from
+    In-Field Photovoltaic I-V Curves on a Single Board Computer, 46th IEEE
+    Photovoltaic Specialist Conference, Chicago, IL, 2019
+    """
+
+    # Find beta0 and beta1 from linear portion of the IV curve
+    beta0, beta1 = _find_beta0_beta1(v, i, vlim, v_oc)
+
+    if not np.isnan(beta0):
+        # Subtract the IV curve from the linear fit. Select points where
+        # beta0 + beta*V - I > ilim * i_sc
+        # in order to find beta3 and beta4 from exponential portion of IV curve
+        y = beta0 - beta1 * v - i
+        x = np.array([np.ones_like(v), v, i]).T
+        beta3, beta4 = _find_beta3_beta4(y, x, ilim, i_sc)
+
+    # calculate single diode parameters from regression coefficients
+    IL, I0, Rsh, Rs, nNsVth = _calculate_sde_parameters(beta0, beta1, beta3,
+                                                        beta4, v_mp, i_mp,
+                                                        v_oc)
+
+    return IL, I0, Rsh, Rs, nNsVth
+
+
+def _calc_I0(IL, I, V, Gp, Rs, beta3):
+    return (IL - I - Gp * V - Gp * Rs * I) / np.exp(beta3 * (V + Rs * I))
+
+
+def _find_beta0_beta1(v, i, vlim, v_oc):
+    # Get intercept and slope of linear portion of IV curve.
+    # Start with V =< vlim * v_oc, extend by adding points until slope is
+    # negative (downward).
+    beta0 = np.nan
+    beta1 = np.nan
+    idx = np.searchsorted(v, vlim * v_oc)
+    while idx <= len(v):
+        coef = np.polyfit(v[:idx], i[:idx], deg=1)
+        if coef[0] < 0:
+            # intercept term
+            beta0 = coef[1].item()
+            # sign change of slope to get positive parameter value
+            beta1 = -coef[0].item()
+            break
+        else:
+            idx += 1
+    return beta0, beta1
+
+
+def _find_beta3_beta4(y, x, ilim, i_sc):
+    idx = np.searchsorted(y, ilim * i_sc) - 1
+    result = np.linalg.lstsq(x[idx:, ], np.log(y[idx:]), rcond=None)
+    coef = result[0]
+    beta3 = coef[1].item()
+    beta4 = coef[2].item()
+    return beta3, beta4
+
+
+def _calculate_sde_parameters(beta0, beta1, beta3, beta4, v_mp, i_mp, v_oc):
+    if not any(np.isnan([beta0, beta1, beta3, beta4])):
+        nNsVth = 1.0 / beta3
+        Rs = beta4 / beta3
+        Gp = beta1 / (1.0 - Rs * beta1)
+        Rsh = 1.0 / Gp
+        IL = (1 + Gp * Rs) * beta0
+        # calculate I0
+        I0_v_mp = _calc_I0(IL, i_mp, v_mp, Gp, Rs, beta3)
+        I0_v_oc = _calc_I0(IL, 0, v_oc, Gp, Rs, beta3)
+        if (I0_v_mp > 0) and (I0_v_oc > 0):
+            I0 = 0.5 * (I0_v_mp + I0_v_oc)
+        elif (I0_v_mp > 0):
+            I0 = I0_v_mp
+        elif (I0_v_oc > 0):
+            I0 = I0_v_oc
+        else:
+            I0 = np.nan
```

Comment:
In this case, should you set all the other parameters to be nan too?

Docstring states: "Returns ``NaN`` for each parameter if the fitting is not successful."